### PR TITLE
Initial support for QNN EP

### PR DESCRIPTION
**Pull request by skyline75489**
Added support for QNN EP and related session options

### CODE REVIEW

File: src/config.cpp [hunk 1]
```diff
@@ -84,6 +84,12 @@ struct SessionOptions_Element : JSON::Element {
       v_.enable_cpu_mem_arena = value;
     else if (name == "enable_mem_pattern")
       v_.enable_mem_pattern = value;
+    else if (name == "disable_cpu_ep_fallback")
```
Comment (skyline75489) [inline src/config.cpp:87]: These options are used in the reference script.

File: src/java/src/main/native/utils.h [hunk 2]
```diff
@@ -30,19 +31,24 @@ bool ThrowIfError(JNIEnv* env, OgaResult* result);
...
   CString(JNIEnv* env, jstring str)
-      : env_{env}, str_{str}, cstr{env->GetStringUTFChars(str, /* isCopy */ nullptr)} {
+      : env_{env}, str_{str}, cstr{env->GetStringUTFChars(str, /* isCopy */ nullptr)}, len_{env->GetStringUTFLength(str)} {
   }
 
   const char* cstr;
 
   operator const char*() const { return cstr; }
 
+  std::string utf8String() {
```
Comment (skyline75489) [inline src/java/src/main/native/utils.h:41]: Strictly speaking, this is Java modified version of UTF8. It'll work for ASCII. But I'm not sure how good it is comparing to "normal" UTF8.

File: src/java/src/main/java/ai/onnxruntime/genai/Model.java [hunk 3]
```diff
@@ -6,6 +6,14 @@
 public final class Model implements AutoCloseable {
   private long nativeHandle;
 
+  public Model(String modelPath, String qnnLibSearchPath) throws GenAIException {
```
Comment (skyline75489) [inline src/java/src/main/java/ai/onnxruntime/genai/Model.java:9]: I don't feel like adding public interfaces like this. But I don't have better ideas. @skottmckay  any thought?

**Review by baijumeswani — APPROVED** [review]

File: src/models/model.cpp [hunk 4]
```diff
@@ -388,6 +422,13 @@ void Model::CreateSessionOptions() {
 
       device_type_ = DeviceType::DML;  // We use a DML allocator for input/output caches, but other tensors will use CPU tensors
 #endif
+    } else if (provider_options.name == "qnn") {
+      std::unordered_map<std::string, std::string> opts;
+      for (auto& option : provider_options.options) {
+        opts.emplace(option.first, option.second);
+      }
+
+      ort_options.AppendExecutionProvider("QNN", opts);
```
Comment (skyline75489) [inline src/models/model.cpp:431]: From ORT doc, currently supported providers includes:

   *   QNN
   *   SNPE
   *   XNNPACK

**Pull request merged**